compiler/rustc_lexer/src/lib.rs

```diff
@@ -758,6 +766,74 @@ impl Cursor<'_> {
         false
     }
 
+    /// Eats the double-quoted string and returns `n_hashes` and an error if encountered.
+    fn guarded_double_quoted_string(&mut self) -> TokenKind {
+        // Create a backup to restore later if this
+        // turns out to not be a guarded literal.
+        let backup = self.clone();
```

Comment:
Instead of creating a backup, you can also just peek one char ahead to see if it is `#` or `"` and return `Pound` in that case. In all other cases, assume it's a double quoted string and parse that. That saves us the infinite lookahead here and leaves it to `rustc_parse::lexer` to handle this in an edition gated way.

Still doesn't make me happy, but the ship on rustc_parse's lexer resetting the lexer with edition rules has already sailed because no one noticed it when that was implemented.